doc/specs/#885 - Terminal Settings Model/Actions Addendum.md

```diff
@@ -0,0 +1,334 @@
+ came from. This allows us to minimize the number of actions we serialize when we write
+ to disk, as opposed to serializing the entire list of actions.
+
+`ActionMap` queries will need to check their parent when they cannot find a matching `Command`
+ in their `_ActionList` (or sooner).
+
+### Modifying Actions
+
+There are several ways a command can be modified:
+- change/remove the key chord
+- change the name
+- change the icon
+- change the action
+
+It is important that these modifications are done through `ActionMap` instead of `Command`.
+ This is to ensure that the `ActionMap` is always aligned with `Command`'s values. Thus,
+ we can add the following functions to `ActionMap`:
+
+ ```c++
+ runtimeclass ActionMap
+{
+    void SetKeyChord(Command cmd, KeyChord keys);
+    void SetName(Command cmd, String name);
+    void SetIcon(Command cmd, String iconPath);
+    void SetAction(Command cmd, ShortcutAction action, IActionArgs actionArgs);
+}
+ ```
+
+`SetKeyChord` will need to make sure to modify the `_KeyMap` and the provided `Command`.
+ If the new key chord was already taken, we also need to update the conflicting `Command`
+ and remove its key chord.
+
+`SetName` will need to make sure to modify the `_NameMap` and the provided `Command`.
+ We also need to iterate through the list of commands and see if the old name is still
+ in use by another command. If that's the case, don't remove `_NameMap[oldName]`,
+ instead retarget it.
+
+`SetIcon` will only need to modify the provided `Command`.
+
+`SetAction` will need to begin by updating the provided `Command`'s `ActionAndArgs`.
+ If the generated name is being used, the name will need to be updated. As with `SetName`,
+ we also need to update `_NameMap` and check if the old name is still in use by another command.
+
+Regarding [Layering Actions](#layering-actions), if the `Command` does not exist in the current layer,
+ but exists in a parent layer, we need to...
+ 1. duplicate it
+ 2. store the duplicate in the current layer
+ 3. make the modification to the duplicate
+
+This ensures that the change propagates post-serialization.
+
+### Unbinding actions
+
+Removing a name or a key chord is currently omitted from this spec because there
+ is no Settings UI use case for it at the moment. This class of scenarios is
+ designed for Command Palette customization.
+
+The only kind of unbinding currently in scope is freeing a key chord such that
+ no action is executed on that key stroke. To do this...
+ 1. On the existing `Action`, set `ShortcutAction = Unbound` and `IActionArgs = nullptr`
+ 2. The associated name and `KeyChord` are mapped to this "unbound" action
+
+In explicitly storing an "unbound" action, we are explicitly saying that this key chord
+ must be passed through and this string must be removed from the command palette.
+
+We specifically need an "unbound" action for serialization. This way, we can output something like this

```

Comment:
Need to update this section. After working on this a bit, it looks like we can get by with `ShortcutAction::Invalid`.

We do need this part written out in a bit more detail because we need to know how to update the `ActionMap` when deserializing the JSON below:
```js
{ "command": "unbound", "keys": "ctrl+c" }
```